---
title: "Introduction to Bulk RNAseq data analysis"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_document:
    toc: yes
    toc_float:
      collapsed: false
      smooth_scroll: true
    css: ../css/boxes.css
    includes:
      in_header: ../css/navbar.html
  pdf_document:
    toc: yes
bibliography: ref.bib
---

# Differential Expression of RNA-seq data

```{r setup, echo=FALSE}
options(tibble.print_max = 4, tibble.print_min = 4, max.print=40, 
        tibble.max_extra_cols=2)
knitr::opts_knit$set(root.dir = here::here("course_files"))
```

After exploring our data and confirming that we are happy with its quality, we can proceed to
test for differentially expressed genes. There are a number of packages to
analyse RNA-Seq data. Most people use
[DESeq2](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)
[@Love2014] or
[edgeR](http://bioconductor.org/packages/release/bioc/html/edgeR.html)
[@Robinson2010; @McCarthy2012]. There is also the option to use the
[limma](https://bioconductor.org/packages/release/bioc/html/limma.html) package
and transform the counts using its `voom` function .They are all equally valid
approaches [@Ritchie2015]. There is an informative and honest blog post
[here](https://mikelove.wordpress.com/2016/09/28/deseq2-or-edger/) by Mike Love,
one of the authors of DESeq2, about deciding which to use.

We will use **DESeq2** for the rest of this practical.


## Load data

In the previous session we read the results from Salmon into R and created a 
`txi` object, which we then saved into an "rds" file. We can now load the txi
from that file to start the differential expression analysis. We will also need
the sample meta data sheet

First load the packages we need.

```{r packages, message = FALSE}
library(DESeq2)
library(tidyverse)
```

Now load the data from the earlier session.

```{r loadData}
# import the txi object created from the Salmon counts
txi <- readRDS("preprocessed/r_objects/txi.rds")

# import the sample info metadata
sample_info <- read_tsv("data/samplesheet_corrected.tsv", col_types = "cccc")
```

It is important to be sure that the order of the samples in rows in the sample 
meta data table matches the order of the columns in the data matrix - `DESeq2`
will **not** check this. If the order does not match you will not be running the
analyses that you think you are.

```{r checkSampleNames}
all(colnames(txi$counts) == sample_info$SampleName)
```

Also, _DESeq2_ likes the columns used for modelling to be encoded as **factors** rather than plain character variables. 
Because of this we will modify those columns and convert them to factors: 

```{r}
sample_info <- sample_info %>%
  mutate(TimePoint = as.factor(TimePoint), 
         Status = as.factor(Status))
```


## _DESeqDataSet_ object

### Design formula

First we need to create a design model formula for our analysis. `DESeq2` will 
use this to generate the model matrix, as we have seen in the linear models 
lecture. 

We have two variables in our experiment: "Status" and "Time Point". 

We will fit two models under two assumptions: no interaction and interaction of
these two factors, however, to demonstrate the how `DESeq2` is used we will start
with a simple model which considers "Status" but ignores "Time Point".  

First, create a variable containing the model using standard R 'formula' syntax.

```{r modelForumla}
status_model <- as.formula(~ Status)
```

What does this look like as a model matrix?

```{r modelMatrix}
model.matrix(status_model, data = sample_info)
```

The intercept has been set automatically to the group in the factor that is
alphabetically first: `Infected`.

It would be nice if `Uninfected` were the base line/intercept. To get R to 
use `Uninfected` as the intercept we need to use a `factor`. Let's set factor 
levels on Status to use `Uninfected` as the intercept.

```{r setFactors}
sample_info <- mutate(sample_info, Status = fct_relevel(Status, "Uninfected"))
model.matrix(status_model, data = sample_info)
```

## Build a _DESeq2DataSet_

We don't actually need to pass `DESeq2` the model matrix, instead we pass it the 
design formula and the `sample_info` it will build the matrix itself.

```{r makedds}
# create the DESeqDataSet object
dds_raw <- DESeqDataSetFromTximport(txi = txi,
                                       colData = sample_info,
                                       design = status_model)
```

When we summarised the counts to gene level, `tximport` also calculated an 
average transcript length for each gene for each sample. For a given gene the
average transcript length may vary between samples if different samples are 
using alternative transcripts. `DESeq2` will incorporate this into its 
"normalisation".

## Filtering

Just as we did in a previous section, we should filter out genes that are uninformative:

```{r}
keep <- rowSums(counts(dds_raw)) > 5
dds <- dds_raw[keep, ]
```


## Differential expression analysis

The main `DESeq2` work flow is carried out in 3 steps:

### `estimateSizeFactors`

First, Calculate the "median ratio" normalisation size factors for each sample 
and adjust for average transcript length on a per gene per sample basis.

```{r commonSizeFactors}
dds <- estimateSizeFactors(dds)
```

Let's have a look at what that did.

`DESeq2` has calculated a normalizsation factor for each gene for each sample.

```{r}
head(normalizationFactors(dds))
```

We can use `plotMA` from `limma` to look at the of these normalisation factors
on data in an MA plot. Let's look at **SRR7657882**, the fifth column, which has
the largest normalisation factors.

```{r}
logcounts <- log2(counts(dds, normalized = FALSE)  + 1)

limma::plotMA(logcounts, array = 5, ylim = c(-5, 5))
abline(h = 0, col = "red")
```

```{r}
log_norm_counts <- log2(counts(dds, normalized = TRUE)  + 1)

limma::plotMA(log_norm_counts, array = 5, ylim = c(-5, 5))
abline(h = 0, col = "red")
```

DESeq2 doesn't actually normalise the counts, it uses raw counts and includes
the normalisation factors in the modeling as an "offset". Please see the DESeq2
documentation if you'd like more details on exactly how they are incorporated 
into the algorithm. For practical purposes we can think of it as a normalisation.

### `estimateDispersions`

Next we need to estimate the dispersion parameters for each gene.

```{r genewiseDispersion}
dds <- estimateDispersions(dds)
```

We can plot all three sets of dispersion estimates. It is particularly important
to do this if you change any of the default parameters for this step.

```{r plotDisp}
plotDispEsts(dds)
```


### `nbinomWaldTest`

Finally, apply Negative Binomial GLM fitting and calculate Wald statistics.

```{r applyGLM}
dds <- nbinomWaldTest(dds)
```

## The `DESeq` command

In practice the 3 steps above can be performed in a single step using the 
`DESeq` wrapper function. Performing the three steps separately is useful if you
wish to alter the default parameters of one or more steps, otherwise the `DESeq`
function is fine.

```{r theShortVersion}
dds <- DESeq(dds)
```

## Extracting results

We can generate a table of differential expression results from the DDS object
using the `results` function of DESeq2.

```{r resultsTable}
res_status <- results(dds, alpha = 0.05)
res_status
```


### Exercise 1

:::exercise 

Now we have made our results table using our simple model, lets have a look at which genes are changing and how many pass our 0.05 threshold for the False Discovery Rate (FDR), which is the adjusted p-value generated by _DESeq2_. 

Consider this code, which is used to count how many genes are "TRUE" for the conditional statement:

```r
sum(res_status$padj < 0.05)
```

Based on that command, answer the following questions:

1. How many genes in total are differentially expressed using FDR < 0.05? Note that the code above is not working entirely. Can you think about how to fix it?
2. Using the same threshold, how many genes are differentially up-regulated?
3. How about down-regulated?

<details><summary>Answer</summary>

**1.**

When we use `sum()` function as given we get `NA` as the result:

```{r}
sum(res_status$padj < 0.05)
```

This is because there are several missing values in the `padj` column. 
We can count how many missing values there are with the following code: 

```{r}
sum(is.na(res_status$padj))
```

So, `r sum(is.na(res_status$padj))` genes had a missing value for `padj`. 
We will come back to this issue in the next section. 

For now, if we want to know how many genes in total pass this threshold, we need to use the `na.rm` option with the `sum()` function: 

```{r}
sum(res_status$padj < 0.05, na.rm = TRUE)
```


**2.**

To investigate how many genes were up-regulated, we can modify our condition to also include genes with a positive log fold change:

```{r}
sum(res_status$padj < 0.05 & res_status$log2FoldChange > 0, 
    na.rm = TRUE)
```

**3.**

Similarly, for down-regulated genes we can look for those with negative log fold change:

```{r}
sum(res_status$padj < 0.05 & res_status$log2FoldChange < 0, 
    na.rm = TRUE)
```

</details>

:::


## Independent filtering

In the previous exercise, we noticed that there were `r sum(is.na(res_status$padj))` genes with a missing value for the adjusted p-value. 
Why is this?
From [`DESeq2` manual](https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#independent-filtering-of-results):

> "The results function of the `DESeq2` package performs independent filtering by
> default using the mean of normalized counts as a filter statistic. A threshold 
> on the filter statistic is found which optimizes the number of adjusted p values
> lower than a [specified] significance level".

The default significance level for independent filtering is `0.1`, however, you
should set this to the FDR cut off you are planning to use. We will use `0.05` -
this was the purpose of the `alpha` argument in the previous command.

Remember that in an earlier section we said that there is no need to pre-filter the genes as DESeq2
will do this through a process it calls 'independent filtering'. The genes 
with `NA` are the ones `DESeq2` has filtered out.


## Additive model

### Exercise 2

:::exercise

So far, we have fitted a simple model considering just "Status", but in reality we want to model the effects of both "Status" and "Time Point".
Let's start with the model with only main effects - an **additive model** with no interaction. 
The main assumption here is that the effects of "Status" and "Time Point" are indepedent.

Starting from `dds`, let's update the model design in our object by running the following:

```{r message=FALSE, warning=FALSE}
# update design formula
design(dds) <- ~ TimePoint + Status
```

After running this step, we need to re-run the model fitting routine with the new design: 

1. Update `dds` by using the `DESeq()` function to re-run the size factor estimation, dispersion estimation and model fitting steps.
2. Extract the default contrast using the `results()` function into a new object called `res_additive`.
  1. What contrast are these results for? Hint: if you have constructed the model correctly, then it should be the same contrast as for the `res_status` model.
  2. How many genes have an adjusted p-value of less than 0.05?

<details><summary>Answer</summary>

After running the given line of code, we can confirm our model is updated:

```{r}
design(dds)
```

We could also look at the model matrix, extracting the design formula and metadata directly from our object:

```{r}
model.matrix(design(dds), data = colData(dds))
```

**1.**

To run our analysis, we need to run the `DESeq()` function to fit this new model and thus update our object:

```{r message=FALSE, warning=FALSE}
dds <- DESeq(dds)
```

**2.**

We can extract the default results contrast using the `results()` function:

```{r}
res_additive <- results(dds, alpha = 0.05)
```

We can take a peak at these results: 

```{r}
head(res_additive)
```

Which indicates our contrast is for "Status Infected vs Uninfected", which is the same contrast as before for the `res_status` model. 
However, this time we took "Time Point" into account in the model, so the actual results are different because our model has more information about our samples. 

</details>

:::


## Results contrasts

The `results` function has returned the results for the contrast "Infected vs
Uninfected". Let's have a look at the model matrix to understand why `DESeq2`
has given us this particular contrast.

```{r viewModelMatrix}
model.matrix(design(dds), data = colData(dds))
```

By default, `results` has returned the contrast encoded by the final column in
the model matrix. `DESeq2` has the command `resultsNames` that allows us to
view the contrasts that are available directly from the DESeq2 object.

```{r resultsNames}
resultsNames(dds)
```

Let's rename `res_additive` so that we know which contrast results it contains.

```{r}
res_infected_vs_uninfected <- res_additive
rm(res_additive)
```

Let's get the top 100 genes by adjusted p-value (i.e. false discovery rate, FDR):

```{r topGenesPvV}
topGenesIvU <- res_infected_vs_uninfected %>%
  # convert to regular data.frame
  as.data.frame() %>%
  # add a column with gene names
  rownames_to_column("GeneID") %>%
  # get the top 100 by FDR
  top_n(100, wt = -padj)

head(topGenesIvU)
```

### Exercise 3

:::exercise

If we want a different contrast we can pass the `results` function the 
**contrast** we want, as given by `resultsNames(dds)`.
The general syntax is: 

```r
results(your_dds_object, contrast = list("name_of_contrast"))
```

1. Retrieve the results for the contrast of d33 versus d11. 
2. How many differentially expressed genes are there at FDR < 0.05?

<details><summary>Answer</summary>

**1.**

We can see the names of the contrasts available for our model:

```{r}
resultsNames(dds)
```

The default contrast retuned earlier by `results()` was `Status_Infected_vs_Uninfected`. 
We can modify this using the `name` option of this function:

```{r}
res_d33_vs_d11 <- results(dds, contrast = list("TimePoint_d33_vs_d11"), alpha = 0.05)

head(res_d33_vs_d11)
```

**2.**

We can use the `sum()` function together with a conditional expression to answer this question:

```{r}
sum(res_d33_vs_d11$padj < 0.05, na.rm = TRUE)
```

There are substantially fewer differentially expressed genes between days than there are between treatment status. 

However, with this additive model we're assuming that all the mice change the same across time points (regardless of their treatment status). 
Should we not expect the changes between the time points to be different depending on their treatment?
We investigate this in then next section.

</details>
:::


## Interaction model 

So far we have modeled gene expression as a function of Status and Time Point 
with an additive model. Realistically, we expect the two factors interact such 
that differences in gene expression between infected and uninfected mice are not 
the same at different time points:

![[**PDF version**](images/AdditiveAndInteractionModels.pdf)](images/AdditiveAndInteractionModels.png)

Let's plot a PCA from `vst` transformed data. 
Can you anticipate if the interaction term will be important?

```{r pcaPlot}
vstcounts <- vst(dds_raw, blind = TRUE)
plotPCA(vstcounts,  intgroup = c("Status", "TimePoint"))
```

In this case we can, from both the PCA and our understanding of the biology, be
fairly certain that the interaction model is the appropriate model to use. 


### Exercise 4

:::exercise

When we looked at the PCA it did seem that an interaction model might be
warranted. Let's fit an interaction model to our data.
 
1. Update the `dds` using a design with an interaction between 
   TimePoint and Status. The model formula should be 

   `~TimePoint + Status + TimePoint:Status`

   where `TimePoint:Status` is the parameter for the interaction between 
   TimePoint and Status.

2. Update the object again by running the `DESeq()` function.
3. Using the `resultsNames()` function, can you explain what each of the coefficients means?
  If it helps, you can also try to draw those coefficients on a schematic plot as we've done earlier. 

<details><summary>Answer</summary>

**1.**

We start by updating our `dds` with a new design formula: 

```{r}
design(dds) <- ~ TimePoint + Status + TimePoint:Status
```

**2.**

We then refit the model again: 

```{r}
dds <- DESeq(dds)
```

**3.**

Here is the output from `resultsNames()`, giving us the coefficients from DESeq2: 

```{r}
resultsNames(dds)
```

For this model, we have: 

- "Intercept" refers to our reference group, which in this case are the mice that are _uninfected_ on _day 11_.
- The first coefficient is named "TimePoint_d33_vs_d11" and refers to the difference between _day 33_ and _day 11_ for _uninfected_ mice (the reference level for Status).
- The second coefficent named "Status_Infected_vs_Uninfected" refers to the difference between _infected_ and _uninfected_ for mice sampled on _day 11_ (the reference level for TimePoint).
- The third coefficient is the _interaction term_ itself, which represents how differently infected and uninfected mice respond across the days. 

</details>
:::

## Interaction model contrasts

If we are settled on using the interaction model, then we need to extract 
our contrasts with reference to this. That is, we can no longer ask the general
question "What is the difference in gene expression between Infected and 
Uninfected?", but must rather ask two quesions:

* "What is the difference in gene expression between Infected and 
Uninfected at 11 days post infection?"   
* "What is the difference in gene expression between Infected and 
Uninfected at 33 days post infection?"    

If we view the `resultsNames` for the interaction model, we can see the intercept
is Uninfected and 11 days post infection:

```{r resNamInteraction}
resultsNames(dds)
```

The main effect `Status_Infected_vs_Uninfected` is therefore the difference
between Infected and Uninfected **at 11 days post infection** (the reference level for _TimePoint_). 

```{r, resIntd11}
res_infected_vs_uninfected_d11 <- results(dds,
                                          contrast = list("Status_Infected_vs_Uninfected"),
                                          alpha = 0.05)
```

To get the results for Infected versus Uninfected at 33 days post infection, 
we would need to add the interaction term `TimePointd33.StatusInfected`.

In the help page for `results` it shows us how to do this with a `contrast` in 
example 3.

```{r resIntd33}
res_infected_vs_uninfected_d33 <- results(dds,
                                          contrast = list(c("Status_Infected_vs_Uninfected", 
                                                            "TimePointd33.StatusInfected")),
                                          alpha=0.05)
```

Number of genes with padj < 0.05 for Test v Control at day 11:

```{r nSigIntd11}
sum(res_infected_vs_uninfected_d11$padj < 0.05, na.rm = TRUE)
```

Number of genes with padj < 0.05 for Test v Control at day 33:

```{r nSigIntd33}
sum(res_infected_vs_uninfected_d33$padj < 0.05, na.rm = TRUE)
```

We can see that there is a strong difference in the effects of infection on 
gene expression between days 11 and 33.

### Exercise 5

:::exercise

Let's investigate the uninfected mice
  
1. Extract the results for d33 v d11 for _Uninfected_ mice. Save them into an object called `res_d33_v_d11_uninfected`.
2. How many genes have an adjusted p-value less than 0.05? Is the results surprising?
3. Extract the results for d33 v d11 for _Infected_ mice. How many genes have an adjusted p-value less than 0.05?
4. Which contrast would you extract if we had this question: "Which genes change differently over time between infected and uninfected mice"? 

<details><summary>Answer</summary>

**1.**

Recall the coefficient names from DESeq2:

```{r}
resultsNames(dds)
```

Because _uninfected_ is the reference level for the "Status" variable, we can extract the results between time points for those mice by using only the term "TimePoint_d33_vs_d11":

```{r}
res_d33_v_d11_uninfected <- results(dds,
                                    contrast = list("TimePoint_d33_vs_d11"),
                                    alpha = 0.05)
```

**2.**

```{r}
table(res_d33_v_d11_uninfected$padj < 0.05)
```

Is this remarkable? 
Maybe not. Do we really expect vast gene expression differences between the 
brains of mice that are slightly older than one another? It is possible that 
there could have been confounding factors, such as changes in enviromental 
conditions such as temperature or feeding regime, that may have effected gene
expression. In which case it was important to set the experiment up with 
control for both time points.

**3.**

Because now we want to investigate the difference between timepoints for the _infected_ mice, we need to add the interaction term to our comparison: 

```{r}
res_d33_v_d11_infected <- results(dds,
                                  contrast = list(c("TimePoint_d33_vs_d11", 
                                                    "TimePointd33.StatusInfected")),
                                  alpha = 0.05)
```

**4.**

This question refers to the interaction term itself, so we could extract it like this: 

```{r}
res_interaction <- results(dds,
                           contrast = list("TimePointd33.StatusInfected"),
                           alpha = 0.05)
```


</details>

:::

## Save the results

Finally save the DESeq object with the fitted interaction model, as well as the two DESeq2 results tables.

```{r saveObjects, eval=FALSE}
dir.create("results/deseq2")
saveRDS(dds, "results/deseq2/dds_interaction_model.rds")
saveRDS(res_infected_vs_uninfected_d11, "results/deseq2/res_infected_vs_uninfected_d11.rds")
saveRDS(res_infected_vs_uninfected_d33, "results/deseq2/res_infected_vs_uninfected_d33.rds")
```


--------------------

# References